pkg/reloader/reloader.go

```diff
@@ -311,12 +349,51 @@ func (r *Reloader) Watch(ctx context.Context) error {
 	}
 }
 
+func normalize(logger log.Logger, inputFile, outputFile string) error {
+	b, err := os.ReadFile(inputFile)
+	if err != nil {
+		return errors.Wrap(err, "read file")
+	}
+
+	// Detect and extract gzipped file.
+	if bytes.Equal(b[0:3], firstGzipBytes) {
+		zr, err := gzip.NewReader(bytes.NewReader(b))
+		if err != nil {
+			return errors.Wrap(err, "create gzip reader")
+		}
+		defer runutil.CloseWithLogOnErr(logger, zr, "gzip reader close")
+
+		b, err = io.ReadAll(zr)
+		if err != nil {
+			return errors.Wrap(err, "read compressed config file")
+		}
+	}
+
+	b, err = expandEnv(b)
+	if err != nil {
+		return errors.Wrap(err, "expand environment variables")
+	}
+
+	tmpFile := outputFile + ".tmp"
+	defer func() {
+		_ = os.Remove(tmpFile)
+	}()
+	if err := os.WriteFile(tmpFile, b, 0644); err != nil {
+		return errors.Wrap(err, "write file")
+	}
+	if err := os.Rename(tmpFile, outputFile); err != nil {
+		return errors.Wrap(err, "rename file")
+	}
+	return nil
+}
+
 // apply triggers Prometheus reload if rules or config changed. If cfgOutputFile is set, we also
 // expand env vars into config file before reloading.
 // Reload is retried in retryInterval until watchInterval.
 func (r *Reloader) apply(ctx context.Context) error {
 	var (
 		cfgHash         []byte
+		subCfgsHash     []byte
 		watchedDirsHash []byte
```

Comment:
Separate hashes for all? 🤔 given normalization can be bit expensive PLUS have side effects (env vars)

Reply:
Do you mean you want a separate hash for each config directory? Since this method is fairly idempotent (in terms of handling state instead of individual events, a la `controller-runtime`), I don't think it makes much of a difference. However, I have made this change. Let me know if this is not what you meant. Thanks!